EVGEN/AliGenExtFile.cxx

```diff
@@ -336,6 +337,90 @@ void AliGenExtFile::Generate()
     }
     ///
 
+    fReader->RewindEvent();
+    AliDebugStream(3) << "Final mother particle assignment" << std::endl;
+    for (int iold = 0; iold < nTracks; iold++) {
+      TParticle* orig_particle = fReader->NextParticle();
+      if (!orig_particle) continue;
+      Bool_t selected = selector.isSelected(iold);
+      AliDebugStream(3) << "Particle " << iold << " with PDG = " << orig_particle->GetPdgCode() << std::endl;
+      AliDebugStream(3) << orig_particle->Px() << " " << orig_particle->Py() << " " << orig_particle->Pz() << " " << orig_particle->Energy() << std::endl;
+      if (!selected) {
+        AliDebugStream(3) << "Particle rejected." << std::endl;
+        continue;
+      }
+      Int_t inew = selector.newId(iold);
+      TParticle* particle = fStack->Particle(inew);
+      if (!particle) {
+        AliErrorStream() << "Could not find particle " << inew << std::endl;
+        continue;
+      }
+      else {
+        AliDebugStream(3) << "New index is " << inew << std::endl;
+      }
+      Int_t iparent1 = selector.newId(orig_particle->GetFirstMother());
+      Int_t iparent2 = selector.newId(orig_particle->GetSecondMother());
+      TParticle* parent1 = nullptr;
+      TParticle* parent2 = nullptr;
+      
+      if (iparent1 >= 0) parent1 = fStack->Particle(iparent1);
+      if (iparent2 >= 0) parent2 = fStack->Particle(iparent2);
+      
+      if (parent1 && !parent2) {
+        // Only the first mother exists, nothing to be done
+        AliDebugStream(3) << "Only first mother exists. Confirm parent 1 with index " << iparent1
+        << " and PDG = " << parent1->GetPdgCode() << std::endl;
+      }
+      else if (parent2 && !parent1) {
+        // Only the second mother exists
+        AliDebugStream(3) << "Only second mother exists. Using mother 2 with index " << iparent2
+        << " and PDG = " << parent2->GetPdgCode() << std::endl;
+        particle->SetFirstMother(iparent2);
+      }
+      else if (parent1 && parent2) {
+        // Both mothers exist. Select the mother with the larger PDG code
```

Comment:
In case of 2 parents, I am wondering whether the choice to retain the mother with the highest PDG code could create some side effects or is immune of troubles...
Clearly one decision has to be made on what to do and this one is likely the least worse, but still.

Looking at the MC scheme of PDG (https://pdg.lbl.gov/2018/reviews/rpp2018-rev-monte-carlo-numbering.pdf), I am trying to hunt for troubles.
I do not find a good counter-example but I can neither grant to have a clear view nor being exhaustive.
My main field of concerns would be about : 
- rescatterings in the hadronic phase in AA (2->2, 3->2, etc)
- diquarks Vs baryons (both ID families O(10^3) )
- leptons+gauge bosons vs hadrons ID (gamma gamma in UPC, gamma A in p-Pb, ...)

I will not hold the merging, and going for a concrete test would be already good to move forward.
Just wanted to discuss here, if the problems are expected to be cured once and for all of if we should expect some later unexpected behaviour hitting us without notice.

Reply:
Hi Antonin, I agree with your concerns, the idea is we have this in master now and in a special tag to test for productions so that we can see how this changes the herwing case and also the EPOS comparing with already existing EPOS.

Reply:
I think you can do "test"

Reply:
Why are we not keeping both mothers?  If a model produces particles through interactions (whatever that may be), then the event record presented to GEANT (or similar) should have that reflected.   Simply removing a mother gives the false impression that the daughter particle is from a decay - which it probably isn't and must not be interpreted that way.  

Nor do I see the point of removing a mother from the tree.   What do we gain from that?  It seems to me that we also run the risk of having mothers that are not attached to an "end-vertex" and are thus considered "final state" and propagated through the experiment.  

I really don't see the point of this at all.  Please enlighten me as to what we hope to achieve with this.

Reply:
Hi Christian, thanks for having a look,
as discussed in https://alice.its.cern.ch/jira/browse/ALIROOT-8653, there is a problem in a recent Herwig production. Since Herwig was used in the past in fast simulations, it was suggested to implement the private modifications that worked then, that are in this commit https://github.com/aiola/alice-fast-simulation/commit/18b78d05cfe7b1674acb673bec7bb8572322437f#diff-0625d73f5f714b7d1e8694970ff7b9a4 from Salvatore, that unfortunately is not anymore ALICE. I am not sure what is the reasoning behind either.
Gustavo

Reply:
Hello Christian, Gustavo, (all)
 looking at Salvatore's code, I started with the idea that it is not possible to attach 2 parents or more to the vertex (AliRoot-side limitation ?  HEPmc-side limitation ?). Based on Christian's input, it seems that this is possible to to attach so, at least on paper.
@mfasDa or somebody from PWG-JE : do you have any memory on the justification for Salvatore's choice ?

Reply:
Hi @maireihpc, (all),
I don't recall  what is the motivation for Salvatore's choice. I think the number of parent particles is an AliRoot simulation. In Herwig usually the hard scatterings have two mother particles, which are partons from the beam particles involved in the scattering - probably for SHERPA it is the same. I think starting from those the particles the rest of the chain has only one mother, which are the clusters (gluons) - pdg code 81 and 82, which can split and then fragment at some point.
Cheers
Markus

Reply:
Hello, we could keep the two mothers. However in AliRoot, the mother->child relation is created on the fly from the child->mother relation and there is unfortunately no guarantee that this works for all MCs. For example in Pythia8 two mothers imo1, imo2 can mean: an interval from [imo1, imo2] or two separated mothers. Same for the daughter indices. AliStack assumes that the daughters are contiguous, which for MCs using two mother indices might not be always the case.